src/Storages/PostgreSQL/MaterializedPostgreSQLConsumer.cpp

```diff
@@ -40,125 +56,160 @@ MaterializedPostgreSQLConsumer::MaterializedPostgreSQLConsumer(
-        storages.emplace(table_name, storage_info);
+        storages.emplace(table_name, StorageData(storage_info, log));
+
+    LOG_TRACE(log, "Starting replication. LSN: {} (last: {}), storages: {}",
+              getLSNValue(current_lsn), getLSNValue(final_lsn), storages.size());
 }
 
 
-MaterializedPostgreSQLConsumer::StorageData::StorageData(const StorageInfo & storage_info)
-    : storage(storage_info.storage), buffer(storage_info.storage->getInMemoryMetadataPtr(), storage_info.attributes)
+MaterializedPostgreSQLConsumer::StorageData::StorageData(const StorageInfo & storage_info, Poco::Logger * log_)
+    : storage(storage_info.storage)
+    , table_description(storage_info.storage->getInMemoryMetadataPtr()->getSampleBlock())
+    , columns_attributes(storage_info.attributes)
+    , column_names(storage_info.storage->getInMemoryMetadataPtr()->getColumns().getNamesOfPhysical())
+    , array_info(createArrayInfos(storage_info.storage->getInMemoryMetadataPtr()->getColumns().getAllPhysical(), table_description))
 {
-    auto table_id = storage_info.storage->getStorageID();
-    LOG_TRACE(&Poco::Logger::get("StorageMaterializedPostgreSQL"),
-              "New buffer for table {}, number of attributes: {}, number if columns: {}, structure: {}",
-              table_id.getNameForLogs(), buffer.attributes.size(), buffer.getColumnsNum(), buffer.description.sample_block.dumpStructure());
-}
+    auto columns_num = table_description.sample_block.columns();
+    /// +2 because of _sign and _version columns
+    if (columns_attributes.size() + 2 != columns_num)
+    {
+        throw Exception(ErrorCodes::LOGICAL_ERROR,
+                        "Columns number mismatch. Attributes: {}, buffer: {}",
+                        columns_attributes.size(), columns_num);
+    }
 
+    LOG_TRACE(log_, "Adding definition for table {}, structure: {}",
+              storage_info.storage->getStorageID().getNameForLogs(),
+              table_description.sample_block.dumpStructure());
+}
 
 MaterializedPostgreSQLConsumer::StorageData::Buffer::Buffer(
-    StorageMetadataPtr storage_metadata, const PostgreSQLTableStructure::Attributes & attributes_)
-    : attributes(attributes_)
+    ColumnsWithTypeAndName && columns_,
+    const ExternalResultDescription & table_description_)
 {
-    const Block sample_block = storage_metadata->getSampleBlock();
-
-    /// Need to clear type, because in description.init() the types are appended
-    description.types.clear();
-    description.init(sample_block);
-
-    columns = description.sample_block.cloneEmptyColumns();
-    const auto & storage_columns = storage_metadata->getColumns().getAllPhysical();
-    auto insert_columns = std::make_shared<ASTExpressionList>();
+    if (columns_.end() != std::find_if(
+            columns_.begin(), columns_.end(),
+            [](const auto & col) { return col.name == "_sign" || col.name == "_version"; }))
+    {
+        throw Exception(ErrorCodes::LOGICAL_ERROR,
```

Comment:
Can't user create a columns with such names somehow?

Reply:
Yes, he can of course, this code checks if they were added during ALTER query after MaterializedPostgreSQL table was created successfully as there. But it is written in doc that these column names are reserved for internal usage.

Reply:
Shouldn't it be another error code then `LOGICAL_ERROR` that indicated bug in out code?

Reply:
Yes, good point.